Autotools: Enable clang tool chain to generate libraries (for #312)

MohammedKhajapasha opened the pull request:
Enable clang tool chain for automake to generate shared libexpat libraries.

Signed-off-by: Mohammed Khajapasha <mohammed.khajapasha@intel.com>

hartwork commented on expat/configure.ac:
```diff
@@ -94,6 +94,9 @@ AC_LANG_POP([C++])
 AS_IF([test "$GCC" = yes],
   [AX_APPEND_LINK_FLAGS([-fno-strict-aliasing],[LDFLAGS])])
 
+AS_IF([test "$GCC" = yes],
```
That looks more like non-clang and not like a check of the linker. What is the idea?

MohammedKhajapasha replied:
Updated LD check for clang linker in new patchset

hartwork commented on expat/configure.ac:
```diff
@@ -94,6 +94,9 @@ AC_LANG_POP([C++])
   [AX_APPEND_LINK_FLAGS([-fno-strict-aliasing],[LDFLAGS])])
 
+AS_IF([test "$GCC" = yes],
+      archive_cmds='$CC -shared $pic_flag -o $lib $libobjs $deplibs $compiler_flags')
```
That's an interesting idea!

I have compared this now to what `configure` puts into `libtool` for `LD=ld`:

```
# LD=ld
\$CC -shared \$pic_flag          \$libobjs \$deplibs \$compiler_flags \$wl-soname \$wl\$soname -o \$lib

# Proposed in this pull request
\$CC -shared \$pic_flag -o \$lib \$libobjs \$deplibs \$compiler_flags
```
I noticed the absense of `\$wl-soname \$wl\$soname` as well as a different location of `-o \$lib`. Is that intended?

I tried former command with `CC=clang LD=clang++` and it worked. My libtool is at 2.4.6 over here if it matters.

MohammedKhajapasha commented:
> That looks pretty good already — thanks!
> 
> I'm not sure if additional `\$wl-soname \$wl\$soname` makes things better or worse. We don't have to add it — you think it's better without? Any comment?

Included host os check to cover $wl-soname \$wl\$soname flags for libraries.

> I tried with `LD="$(which clang++)"` now to confirm that pattern `clang*` doesn't cover that case. I changed the pattern to `*clang*` locally with asterisks on both ends, which made things work. Any objections to using `*clang*` for a pattern, instead?

currently the issue associated with clang linker so, as per understanding it is not good with \*clang\* command.

> Addition of a simple comment along the lines of…
> 
> ```
> # We're patching ${archive_cmds} to affect generation of file "libtool" to fix linking with clang (issue #312)
> ```
> 
> …would be great to have.

Updated the commit and subject as above note.

> I think the last comma in the call to `AS_CASE` can be dropped (but doesn't matter much).

Removed last comma in new patchset.

> Will be a cool last addition to go into the upcoming release.

hartwork commented:
> Included host os check to cover $wl-soname $wl$soname flags for libraries.

I see, seems to be related to the macOS trouble I saw at https://travis-ci.org/libexpat/libexpat/jobs/583007576#L1160 .

> > I tried with `LD="$(which clang++)"` now to confirm that pattern `clang*` doesn't cover that case. I changed the pattern to `*clang*` locally with asterisks on both ends, which made things work. Any objections to using `*clang*` for a pattern, instead?
> 
> currently the issue associated with clang linker so, as per understanding it is not good with *clang* command.

I don't understand: `LD=/usr/lib/llvm/10/bin/clang++` is clang; it is not matched by pattern `clang*` (i.e. build still broken) but is matched by `*clang*` (i.e. build succeeds). If pattern `*clang*` is too greedy, please help me understand why.

> > Addition of a simple comment along the lines of…
> > ```
> > # We're patching ${archive_cmds} to affect generation of file "libtool" to fix linking with clang (issue #312)
> > ```
> > 
> > 
> > …would be great to have.
> 
> Updated the commit and subject as above note.

I think we have a misunderstanding here. I was suggesting adding that comment right into `configure.ac`. The commit message is good, except that we're not patching `automake` here.

The second `[AS_CASE` does not match the `configure.ac` style guide we introduced in #315 with @kkkunche, recently. It's the third parameter the first `AS_CASE`, so https://github.com/libexpat/libexpat/blob/master/expat/configure-ac-style.md#2-parameter-indentation applies.

I'm happy to add the comment and adjust the formatting myself but wanted you to know — as you like.

MohammedKhajapasha commented:
> > Included host os check to cover $wl-soname $wl$soname flags for libraries.
> 
> I see, seems to be related to the macOS trouble I saw at https://travis-ci.org/libexpat/libexpat/jobs/583007576#L1160 .
> 
> > > I tried with `LD="$(which clang++)"` now to confirm that pattern `clang*` doesn't cover that case. I changed the pattern to `*clang*` locally with asterisks on both ends, which made things work. Any objections to using `*clang*` for a pattern, instead?
> > 
> > 
> > currently the issue associated with clang linker so, as per understanding it is not good with _clang_ command.
> 
> I don't understand: `LD=/usr/lib/llvm/10/bin/clang++` is clang; it is not matched by pattern `clang*` (i.e. build still broken) but is matched by `*clang*` (i.e. build succeeds). If pattern `*clang*` is too greedy, please help me understand why.

Currently we don't have any setup with macOS, will grep and update on this. Is there any prefix used for clang in macOS case (like macos_clang.....etc ) ?

> > > Addition of a simple comment along the lines of…
> > > ```
> > > # We're patching ${archive_cmds} to affect generation of file "libtool" to fix linking with clang (issue #312)
> > > ```
> > > 
> > > 
> > > …would be great to have.
> > 
> > 
> > Updated the commit and subject as above note.
> 
> I think we have a misunderstanding here. I was suggesting adding that comment right into `configure.ac`. The commit message is good, except that we're not patching `automake` here.

Taken care in new patchset.

> The second `[AS_CASE` does not match the `configure.ac` style guide we introduced in #315 with @kkkunche, recently. It's the third parameter the first `AS_CASE`, so https://github.com/libexpat/libexpat/blob/master/expat/configure-ac-style.md#2-parameter-indentation applies.

Taken care in new patchset.

> I'm happy to add the comment and adjust the formatting myself but wanted you to know — as you like.

hartwork commented:
Excellent, thank you!
